tests/c/udp-push-pop.c

```diff
@@ -0,0 +1,238 @@
+ *====================================================================================================================*/
+
+/**
+ * @brief UDP server.
+ *
+ * @param argc   Argument count.
+ * @param argv   Argument list.
+ * @param local  Local socket address.
+ */
+static void server(int argc, char *const argv[], struct sockaddr_in *local)
+{
+    int sockqd = -1;
+
+    /* Initialize demikernel */
+    assert(demi_init(argc, argv) == 0);
+
+    /* Setup peer. */
+    assert(demi_socket(&sockqd, AF_INET, SOCK_DGRAM, 0) == 0);
+    assert(demi_bind(sockqd, (const struct sockaddr *)local, sizeof(struct sockaddr_in)) == 0);
+
+    /* Run peer. */
+    for (int it = 0; it < MAX_ITERATIONS; it++)
+    {
+        demi_qtoken_t qt = -1;
+        demi_qresult_t qr = {0};
+
+        /* Push data. */
+        assert(demi_pop(&qt, sockqd) == 0);
+
+        /* Wait push operation to complete. */
+        assert(demi_wait(&qr, qt) == 0);
+
+        /* Parse operation result. */
+        assert(qr.qr_opcode == DEMI_OPC_POP);
+        assert(qr.qr_value.sga.sga_segs != 0);
+
+        /* Release scatter-gather array. */
+        assert(demi_sgafree(&qr.qr_value.sga) == 0);
+
+        fprintf(stdout, "pop (%d)\n", it);
+    }
+}
+
+/*====================================================================================================================*
+ * client()                                                                                                           *
+ *====================================================================================================================*/
+
+/**
+ * @brief UDP client.
+ *
+ * @param argc   Argument count.
+ * @param argv   Argument list.
+ * @param local  Local socket address.
+ * @param remote Remote socket address.
+ */
+static void client(int argc, char *const argv[], struct sockaddr_in *local, struct sockaddr_in *remote)
+{
+    int sockqd = -1;
+
+    /* Initialize demikernel */
+    assert(demi_init(argc, argv) == 0);
+
+    /* Setup peer. */
+    assert(demi_socket(&sockqd, AF_INET, SOCK_DGRAM, 0) == 0);
+    assert(demi_bind(sockqd, (const struct sockaddr *)local, sizeof(struct sockaddr_in)) == 0);
+
+    /* Run peer. */
+    for (int it = 0; it < MAX_ITERATIONS; it++)
+    {
+        demi_qtoken_t qt = -1;
+        demi_qresult_t qr = {0};
+        demi_sgarray_t sga = {0};
+
+        /* Allocate scatter-gather array. */
+        sga = demi_sgaalloc(DATA_SIZE);
+        assert(sga.sga_segs != 0);
+
+        /* Cook data. */
+        memset(sga.sga_segs[0].sgaseg_buf, (it & 0xff), DATA_SIZE);
+
+        /* Push data. */
+        assert(demi_pushto(&qt, sockqd, &sga, (const struct sockaddr *)remote, sizeof(struct sockaddr_in)) == 0);
+
+        /* Wait push operation to complete. */
+        assert(demi_wait(&qr, qt) == 0);
+
+        /* Parse operation result. */
+        assert(qr.qr_opcode == DEMI_OPC_PUSH);
+
+        /* Release scatter-gather array. */
+        assert(demi_sgafree(&sga) == 0);
```

Comment:
This comment depends upon what exactly you are wanting to test with this test.  But since you wait for the push operation to complete before you attempt another, why not just allocate the buffer once before the loop and wait to free it after you exit the loop?

Reply:
- This test should assess the same scenario than the one written in rust.
  - This test should exercise multiple SGAs allocations
  - This test should not perform in-place modifications